core/state-machine/src/lib.rs

```diff
@@ -639,49 +330,57 @@ impl<'a, H, N, B, T, O, Exec> StateMachine<'a, H, N, B, T, O, Exec> where
 		NC: FnOnce() -> result::Result<R, &'static str> + UnwindSafe,
 		Handler: FnOnce(
 			CallResult<R, Exec::Error>,
-			CallResult<R, Exec::Error>
+			CallResult<R, Exec::Error>,
 		) -> CallResult<R, Exec::Error>
 	{
 		// read changes trie configuration. The reason why we're doing it here instead of the
 		// `OverlayedChanges` constructor is that we need proofs for this read as a part of
 		// proof-of-execution on light clients. And the proof is recorded by the backend which
 		// is created after OverlayedChanges
 
-		let backend = self.backend.clone();
-		let init_overlay = |overlay: &mut OverlayedChanges, final_check: bool| {
+		let init_overlay = |overlay: &mut OverlayedChanges, final_check: bool, backend: &B| {
 			let changes_trie_config = try_read_overlay_value(
 				overlay,
 				backend,
 				well_known_keys::CHANGES_TRIE_CONFIG
 			)?;
 			set_changes_trie_config(overlay, changes_trie_config, final_check)
 		};
-		init_overlay(self.overlay, false)?;
+		init_overlay(self.overlay, false, &self.backend)?;
 
 		let result = {
 			let orig_prospective = self.overlay.prospective.clone();
 
 			let (result, storage_delta, changes_delta) = match manager {
 				ExecutionManager::Both(on_consensus_failure) => {
-					self.execute_call_with_both_strategy(compute_tx, native_call.take(), orig_prospective, on_consensus_failure)
+					self.execute_call_with_both_strategy(
+						compute_tx,
+						native_call.take(),
+						orig_prospective,
+						on_consensus_failure,
+					)
 				},
 				ExecutionManager::NativeElseWasm => {
-					self.execute_call_with_native_else_wasm_strategy(compute_tx, native_call.take(), orig_prospective)
+					self.execute_call_with_native_else_wasm_strategy(
+						compute_tx,
+						native_call.take(),
+						orig_prospective,
+					)
 				},
 				ExecutionManager::AlwaysWasm => {
-					let (result, _, storage_delta, changes_delta) = self.execute_aux(compute_tx, false, native_call);
-					(result, storage_delta, changes_delta)
+					let res = self.execute_aux(compute_tx, false, native_call);
+					(res.0, res.2, res.3)
 				},
 				ExecutionManager::NativeWhenPossible => {
-					let (result, _was_native, storage_delta, changes_delta) = self.execute_aux(compute_tx, true, native_call);
-					(result, storage_delta, changes_delta)
+					let res = self.execute_aux(compute_tx, true, native_call);
+					(res.0, res.2, res.3)
```

Comment:
Doesn't look like improvement to me. Shouldn't we just introduce a nominal type for this?

Reply:
Not sure this needs to be done in this PR. We can also put this into a nice starter-ticket.

Reply:
Good idea1